### PR TITLE
Structural/typescript/decorator

### PR DESCRIPTION
**Pull request by nickjedwards**

### CODE REVIEW

File: exercise-plans/structural/decorator.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+## Intent
+
+**Decorator** is a structural design pattern that lets you attach new behaviors to objects by placing these objects
+inside special wrapper objects that contain the behaviors.
+
+## Problem
+
+We have a small but growing coffee shop that currently has some pretty annoying customers in the controller code.
+At the moment we're having to create concrete classes for every possible combonation of coffee order and instead,
```
Comment (cdaloisio) [inline exercise-plans/structural/decorator.md]:
```suggestion
At the moment we're having to create concrete classes for every possible combination of coffee order and instead,
```

File: typescript/structural/decorator/models/Coffee.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+export default interface Coffee {  
+  cost(): number
```
Comment (cdaloisio) [inline typescript/structural/decorator/models/Coffee.ts:10]: Given that cost might be something that has to be separated or factored out per decorator, could you add a comment here for the price list of the individual items. ie:
```
// # Price List
// InstantCoffee - 1.0
// RoastedCoffee - 1.2
// Milk - 0.5
// Sugar - 0.25
```

File: typescript/structural/decorator/controllers/CoffeeController.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    // * Identify the primary component that has optional layers
+    // * Identify the methods in common to both the primary component and the optional layers
+    // * Create the component interface and declare the common methods
+    // * Create the concrete component class and define the base behavior
+    // * Create the base decorator class. It should have a property for referencing the base or decorated object
+    // * Create concrete decorators by extending them from the base decorator
+    // * Compose!
+
+    // A revolutionary customer decides they want Milk and Sugar with their coffee... wuuuttt?
+    // ...Should we create a new concrete class InstantCoffeeWithMilkAndSugar? Ah ah ah
```
Comment (cdaloisio) [inline typescript/structural/decorator/controllers/CoffeeController.ts:18]: This is great as a next step - are you expecting that people would have already refactored the code at this point? If so, the reference to the concrete classes would have been changed already. If you want to highlight the point though, perhaps move this before the TODO above. Same with the other one below.

**Review by cdaloisio — CHANGES_REQUESTED** [review]
Love it @nickjedwards  :+1: :100: Just a couple of comments/suggestions before merging and should be good to go!

File: typescript/structural/decorator/controllers/CoffeeController.ts (hunk 3 above)
Comment (nickjedwards) [reply typescript/structural/decorator/controllers/CoffeeController.ts:18]: Yes, I'm expecting they would have created the necessary interfaces and abstraction. When trialing my own task, for me I left those concretions hanging around and just composed the next two "coffee orders".
I like your approach, I think it's best practice to also refactor the existing code (I would in any given real project!), so why not!

Comment (cdaloisio) [reply typescript/structural/decorator/controllers/CoffeeController.ts:18]: Love it :smile:

**Review by cdaloisio — APPROVED** [review]
Brilliant - I'll get this merged now :+1: :rocket: :shipit:

**Pull request merged**